deepchem/models/tests/test_losses.py

```diff
@@ -0,0 +1,199 @@
+  has_tensorflow = False
+
+try:
+  import torch
+  has_pytorch = True
+except:
+  has_pytorch = False
+
+
+class TestLosses(unittest.TestCase):
+  """Test loss functions."""
+
+  @unittest.skipIf(not has_tensorflow, 'TensorFlow is not installed')
+  def test_l1_loss_tf(self):
+    """Test L1Loss."""
+    loss = losses.L1Loss()
+    outputs = tf.constant([[0.1, 0.8], [0.4, 0.6]])
+    labels = tf.constant([[0.0, 1.0], [1.0, 0.0]])
+    result = loss._compute_tf_loss(outputs, labels).numpy()
+    expected = [[0.1, 0.2], [0.6, 0.6]]
+    assert np.allclose(expected, result)
+
+  @unittest.skipIf(not has_pytorch, 'PyTorch is not installed')
+  def test_l1_loss_pytorch(self):
+    """Test L1Loss."""
+    loss = losses.L1Loss()
+    outputs = torch.tensor([[0.1, 0.8], [0.4, 0.6]])
+    labels = torch.tensor([[0.0, 1.0], [1.0, 0.0]])
+    result = loss._create_pytorch_loss()(outputs, labels).numpy()
+    expected = [[0.1, 0.2], [0.6, 0.6]]
+    assert np.allclose(expected, result)
+
+  @unittest.skipIf(not has_tensorflow, 'TensorFlow is not installed')
+  def test_l2_loss_tf(self):
+    """Test L2Loss."""
+    loss = losses.L2Loss()
+    outputs = tf.constant([[0.1, 0.8], [0.4, 0.6]])
+    labels = tf.constant([[0.0, 1.0], [1.0, 0.0]])
+    result = loss._compute_tf_loss(outputs, labels).numpy()
+    expected = [[0.1**2, 0.2**2], [0.6**2, 0.6**2]]
+    assert np.allclose(expected, result)
+
+  @unittest.skipIf(not has_pytorch, 'PyTorch is not installed')
+  def test_l2_loss_pytorch(self):
+    """Test L2Loss."""
+    loss = losses.L2Loss()
+    outputs = torch.tensor([[0.1, 0.8], [0.4, 0.6]])
+    labels = torch.tensor([[0.0, 1.0], [1.0, 0.0]])
+    result = loss._create_pytorch_loss()(outputs, labels).numpy()
+    expected = [[0.1**2, 0.2**2], [0.6**2, 0.6**2]]
+    assert np.allclose(expected, result)
+
+  @unittest.skipIf(not has_tensorflow, 'TensorFlow is not installed')
+  def test_hinge_loss_tf(self):
+    """Test HingeLoss."""
+    loss = losses.HingeLoss()
+    outputs = tf.constant([[0.1, 0.8], [0.4, 0.6]])
+    labels = tf.constant([[1.0, -1.0], [-1.0, 1.0]])
+    result = loss._compute_tf_loss(outputs, labels).numpy()
+    expected = [np.mean([0.9, 1.8]), np.mean([1.4, 0.4])]
+    assert np.allclose(expected, result)
+
+  @unittest.skipIf(not has_pytorch, 'PyTorch is not installed')
+  def test_hinge_loss_pytorch(self):
+    """Test HingeLoss."""
+    loss = losses.HingeLoss()
+    outputs = torch.tensor([[0.1, 0.8], [0.4, 0.6]])
+    labels = torch.tensor([[1.0, -1.0], [-1.0, 1.0]])
+    result = loss._create_pytorch_loss()(outputs, labels).numpy()
+    expected = [np.mean([0.9, 1.8]), np.mean([1.4, 0.4])]
+    assert np.allclose(expected, result)
+
+  @unittest.skipIf(not has_tensorflow, 'TensorFlow is not installed')
+  def test_binary_cross_entropy_tf(self):
+    """Test BinaryCrossEntropy."""
+    loss = losses.BinaryCrossEntropy()
+    outputs = tf.constant([[0.1, 0.8], [0.4, 0.6]])
+    labels = torch.tensor([[0.0, 1.0], [1.0, 0.0]])
```

Comment:
Why?  It works fine.

```
>>> torch.tensor([[0.0, 1.0], [1.0, 0.0]])
tensor([[0., 1.],
        [1., 0.]])
```

Reply:
In general, do we calculate loss values between torch.Tensor and tf.Tensor? I think this answer is No. If we use tensorflow, labels and outputs are also defined by tf.Tensor. So, I think it is better to define labels using tf.Tensor compared with torch.Tensor. 

And, the test name is `test_binary_cross_entropy_tf`.
I seem that `XXX_tf` tests expect only tensorflow dependency. (`unittest.skipIf` also implify this) So, I think it is not better to depend on PyTorch.  (If you think you really want to use PyTorch in order to define labels using torch.Tensor, I think you should also add `@unittest.skipIf(not has_pytorch, 'PyTorch is not installed')`)

Reply:
Oh, I see!  I though you were saying it should be `torch.constant()` instead of `torch.tensor()`.  Yes, this was an accident.  TensorFlow tests should use TensorFlow classes and PyTorch tests should use PyTorch classes.